### DISCUSSION TITLE
How to get past hCaptcha asking to select similar images ?

### DISCUSSION THREAD

**Discussion by vivekthoppil** [seleniumbase/SeleniumBase#2892] [Q&A] [answered]
While implementing an automated screenshot taking logic for a Url, the site is asking for selecting similar images possibly due to the detection of bot by hCaptcha . I am using `uc` mode.

 **Manually navigating to the Url in the browser is working.** 

Code

```python
    from seleniumbase import SB
    fallback_ua = 'Mozilla/5.0 (Linux; Android 10; K) AppleWebKit/537.36 (KHTML, like Gecko) Chrome/114.0.0.0 Mobile Safari/537.36'

    chromium_arg = '--disable-gpu,--disable-blink-features=AutomationControlled'
    with SB(uc=True, headless2=True, agent=fallback_ua, chromium_arg=chromium_arg) as sb:
        url = "https://www.royalmail.com/track-your-item#/tracking-results/QF085212272GB"
        sb.uc_open_with_reconnect(url, 3)
        sb.save_screenshot('myimage.png')
```

**Result**

![myimage (1)](https://github.com/seleniumbase/SeleniumBase/assets/30136421/1a381965-c437-481f-9a44-87ec59aaa16b)

_Dockerized environment_: python3.9-slim

_seleniumbase_: 4.28.0

_Chromium Browser_: Chromium 126.0.6478.126 built on Debian 12.5, running on Debian 12.5

_Chrome Driver_: ChromeDriver 126.0.6478.126 (d36ace6122e0a59570e258d82441395206d60e1c-refs/branch-heads/6478@{#1591})

**Comment by mdmintz** [comment, marked as answer]
It looks like you're changing things that you shouldn't be. Use this instead:

```python
from seleniumbase import SB

with SB(uc=True) as sb:
    url = "https://www.royalmail.com/track-your-item#/tracking-results/QF085212272GB"
    sb.uc_open_with_reconnect(url, 4)
    sb.save_screenshot("myimage.png")
```

Don't use headless mode, don't change the default user-agent, and don't set chromium args in UC Mode.

**Reply by vivekthoppil** [reply]
I am running the code in a debian based docker container. If i remove headless2 , getting the below exception

> selenium.common.exceptions.SessionNotCreatedException: Message: session not created: cannot connect to chrome at 127.0.0.1:9222
from chrome not reachable

**Reply by mdmintz** [reply]
Try with `4.28.3` (or newer if available).

Also, a Docker environment is much more detectable than regular Linux.

**Reply by vivekthoppil** [reply]
@mdmintz  I ran it using 4.28.3 directly on my MacOs and it is working. However i have some follow up questions. It would be very helpful if you could answer those.

1) I plan to deploy my application in heroku where there will be no GUI. And hence i want the headless mode. How can i configure that ?
2) Basically i already have a working model using the raw `undetected_chrome_driver` and want to convert the whole project to be based on SeleniumBase. Currently ( before conversion ) , i use the below chromium args . Can u tell me if those are reduntant ? If not, how can i configure those in SB ? Note: I plan to use bot `UC` mode non UC mode depending on the site.

```
    import undetected_chromedriver as uc

    options = uc.ChromeOptions()
    options.headless=True
    
    options.add_argument('--headless')
    options.add_argument(f'--user-agent=Mozilla/5.0 (Linux; Android 10; K) AppleWebKit/537.36 (KHTML, like Gecko) Chrome/114.0.0.0 Mobile Safari/537.36')
    options.add_argument("start-maximized")
    options.add_argument('--no-sandbox')
    options.add_argument('--disable-dev-shm-usage')
    options.add_argument("--disable-gpu")
    options.add_argument("--window-size=1920,1080")
    options.add_argument("--disable-blink-features=AutomationControlled")
    options.add_experimental_option(
    "prefs", {"useAutomationExtension": False, "excludeSwitches": ["enable-automation"]})
    driver = uc.Chrome(options=options)
    driver.execute_script("Object.defineProperty(navigator, 'webdriver', {get: () => undefined})")
```

**Reply by mdmintz** [reply]
No GUI is not a problem. SeleniumBase sets up an X11 Xvfb virtual display so that headless mode isn't needed to run in a GUI-less environment.

Don't set any of those args in UC Mode because UC Mode already sets them. This isn't the original `undetected_chromedriver`. SeleniumBase already sets all the important args for you.

**Reply by vivekthoppil** [reply]
I deployed a Flask app in heroku which returns the screenshot of a url.

```python
@app.route('/test')
def test():
    from seleniumbase import SB
    with SB(uc=True) as sb:
        url = "https://www.royalmail.com/track-your-item#/tracking-results/QF085212272GB"
        sb.uc_open_with_reconnect(url, 4)
        sb.save_screenshot('myimage.png')
    
        return send_file('./myimage.png', mimetype='image/png')
```

The above code gave an error

> selenium.common.exceptions.SessionNotCreatedException: Message: session not created: cannot connect to chrome at 127.0.0.1:9222

The log contained  ( not the error log ) 

> X11 display failed! Will use regular xvfb!

I tried your [comment here](https://github.com/seleniumbase/SeleniumBase/discussions/2426#discussioncomment-8112765) and put `xvfb=True` ( which did not work ) followed by `headless2=True` . The issue with chrome is now over, **but now the SeleniumBase is getting detected by hCaptcha , which is the original issue.** 

**The heroku-22 stack, which is the environment for the app is based on Ubuntu 22.04**

**Reply by mdmintz** [reply]
It sounds like the Linux environment you used does not support the required X11 display.

To try launching the X11 display on your own, set `pyautogui._pyautogui_x11._display` to `Xlib.display.Display(os.environ['DISPLAY'])` on Linux in order to sync up `pyautogui` with the `X11` virtual display after calling `sbvirtualdisplay.Display(visible=True, size=(1366, 768), backend="xvfb", use_xauth=True).start()`. (For `Xlib`, use `import Xlib.display` after `pip install python-xlib`.)

**Reply by vivekthoppil** [reply]
What exactly is `os.environ['DISPLAY']` ? I dont see any env variables set in that name .

I am getting

> 2024-07-02T13:40:49.795786+00:00 app[web.1]: _display = Display(os.environ['DISPLAY'])
2024-07-02T13:40:49.795786+00:00 app[web.1]: ~~~~~~~~~~^^^^^^^^^^^
2024-07-02T13:40:49.795786+00:00 app[web.1]: File "<frozen os>", line 714, in __getitem__
2024-07-02T13:40:49.795786+00:00 app[web.1]: KeyError: 'DISPLAY'

**Code**

```python
from sbvirtualdisplay import Display
import Xlib.display
import pyautogui

    try:
        disp = Display(visible=True, size=(1366, 768), backend="xvfb", use_xauth=True)
        disp.start()
        pyautogui._pyautogui_x11._display = Xlib.display.Display(os.environ['DISPLAY'])
        from seleniumbase import SB
        with SB(uc=True) as sb:
            sb.uc_open_with_reconnect(url, 4)
            sb.save_screenshot('myimage.png')
            return send_file('./myimage.png', mimetype='image/png')
    finally:
        disp.stop()
```

**Reply by mdmintz** [reply]
It is automatically set when running an X11 / Xvfb display on Linux. If the variable isn't there, it means the display didn't start. See https://github.com/ponty/PyVirtualDisplay for more info. SeleniumBase uses a modified fork called `sbvirtualdisplay`, but 90% of the code is the same.

**Reply by vivekthoppil** [reply]
I installed the below libraries

```
xvfb 
xauth
python3-tk
```

and ran the below code in both Docker environment and Heroku environment

```python
    from seleniumbase import SB
    from sbvirtualdisplay import Display
    import Xlib.display

    disp = Display(visible=True, size=(1920, 1080), backend="xvfb", use_xauth=True)
    disp.start()
    try:
        import pyautogui
        pyautogui._pyautogui_x11._display = Xlib.display.Display(os.environ['DISPLAY'])
        with SB(uc=True) as sb:
            url = "https://www.royalmail.com/track-your-item#/tracking-results/QF085212272GB"
            sb.uc_open_with_reconnect(url, 4)
            sb.save_screenshot('myimage.png')
            return send_file('./myimage.png', mimetype='image/png')
    finally:
        disp.stop()
```

**Docker** -- > The program ran without error but the hCaptcha issue still persists. Something to do with docker ?

**Heroku** -- > Program did not run. Below exception is happening.

```
b.1]: Traceback (most recent call last):
2024-07-03T11:25:41.800524+00:00 app[web.1]:   File "/app/app.py", line 17, in test
2024-07-03T11:25:41.800525+00:00 app[web.1]:     import pyautogui
2024-07-03T11:25:41.800525+00:00 app[web.1]:   File "/app/.heroku/python/lib/python3.12/site-packages/pyautogui/__init__.py", line 246, in <module>
2024-07-03T11:25:41.800525+00:00 app[web.1]:     import mouseinfo
2024-07-03T11:25:41.800525+00:00 app[web.1]:   File "/app/.heroku/python/lib/python3.12/site-packages/mouseinfo/__init__.py", line 223, in <module>
2024-07-03T11:25:41.800525+00:00 app[web.1]:     _display = Display(os.environ['DISPLAY'])
2024-07-03T11:25:41.800525+00:00 app[web.1]:                ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
2024-07-03T11:25:41.800526+00:00 app[web.1]:   File "/app/.heroku/python/lib/python3.12/site-packages/Xlib/display.py", line 89, in __init__
2024-07-03T11:25:41.800526+00:00 app[web.1]:     self.display = _BaseDisplay(display)
2024-07-03T11:25:41.800526+00:00 app[web.1]:                    ^^^^^^^^^^^^^^^^^^^^^
2024-07-03T11:25:41.800527+00:00 app[web.1]:   File "/app/.heroku/python/lib/python3.12/site-packages/Xlib/display.py", line 71, in __init__
2024-07-03T11:25:41.800527+00:00 app[web.1]:     protocol_display.Display.__init__(self, *args, **keys)
2024-07-03T11:25:41.800527+00:00 app[web.1]:   File "/app/.heroku/python/lib/python3.12/site-packages/Xlib/protocol/display.py", line 89, in __init__
2024-07-03T11:25:41.800527+00:00 app[web.1]:     self.socket = connect.get_socket(name, protocol, host, displayno)
2024-07-03T11:25:41.800527+00:00 app[web.1]:                   ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
2024-07-03T11:25:41.800528+00:00 app[web.1]:   File "/app/.heroku/python/lib/python3.12/site-packages/Xlib/support/connect.py", line 87, in get_socket
2024-07-03T11:25:41.800528+00:00 app[web.1]:     return mod.get_socket(dname, protocol, host, dno)
2024-07-03T11:25:41.800528+00:00 app[web.1]:            ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
2024-07-03T11:25:41.800528+00:00 app[web.1]:   File "/app/.heroku/python/lib/python3.12/site-packages/Xlib/support/unix_connect.py", line 112, in get_socket
2024-07-03T11:25:41.800528+00:00 app[web.1]:     raise error.DisplayConnectionError(dname, str(val))
2024-07-03T11:25:41.800528+00:00 app[web.1]: Xlib.error.DisplayConnectionError: Can't connect to display ":1001": [Errno 111] Connection refused
```

Any idea why this is happening ?

Basically i just need to run SB in Linux environment.
